components/brave_shields/browser/tracking_protection_helper.cc

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2019 The Brave Authors. All rights reserved.
+
+#include "brave/components/brave_shields/browser/tracking_protection_helper.h"
+
+#include "brave/browser/brave_browser_process_impl.h"
+#include "brave/components/brave_shields/browser/tracking_protection_service.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/render_process_host.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+using content::NavigationHandle;
+using content::RenderFrameHost;
+using content::WebContents;
+
+namespace brave_shields {
+
+TrackingProtectionHelper::TrackingProtectionHelper(WebContents*
+  web_contents)
+    : WebContentsObserver(web_contents) {
+}
+
+TrackingProtectionHelper::~TrackingProtectionHelper() {
+}
+
+void TrackingProtectionHelper::DidStartNavigation(NavigationHandle* handle) {
+  RenderFrameHost* rfh = web_contents()->GetMainFrame();
+
+  if (!ui::PageTransitionIsRedirect(handle->GetPageTransition())) {
+    g_brave_browser_process->tracking_protection_service()->
+      SetStartingSiteForRenderFrame(handle->GetURL(),
```

Comment:
Maybe add a comment and/or link to the design doc describing this in more detail?
I'm not sure I got the idea of this change. Image, the user visits 'instagram.com'. There are two subframes that programmaticaly navigate to facebook domains (we will not save 'starting sites' because of `ui::PageTransitionIsRedirect`) . 

Should we block tracking requests from these frames?

Do we also block tracking requests from service workers? Their requests to IdexedDB, are not related to navigations at all.

Reply:
Also, correct me if I'm wrong: after firing `DidStartNavigation`, a navigation could be canceled. But we already updated the host in the service, thus subsequent tracking requests (that can be originated from previous host, since the navigation to a new one is aborted) could be blocked.

Reply:
> Maybe add a comment and/or link to the design doc describing this in more detail?
Here is the link to the spec: https://docs.google.com/document/d/1jolVinFcN4sl0z1zdIHyGdV594Yz6RY4zssnm_VJOYg/edit#. I included it in the PR description as well

> I'm not sure I got the idea of this change. Image, the user visits 'instagram.com'. There are two subframes that programmaticaly navigate to facebook domains (we will not save 'starting sites' because of `ui::PageTransitionIsRedirect`) .
> 
> Should we block tracking requests from these frames?
> 
We are only blocking requests for top-level navigations. When RenderFrameHostChanged is invoked we check if the old/new host are parent frames.

> Do we also block tracking requests from service workers? Their requests to IdexedDB, are not related to navigations at all.
No

Reply:
> Also, correct me if I'm wrong: after firing `DidStartNavigation`, a navigation could be canceled. But we already updated the host in the service, thus subsequent tracking requests (that can be originated from previous host, since the navigation to a new one is aborted) could be blocked.

We only set the host as the `starting site` if the navigation was *NOT* a because of a redirect. If the navigation is cancelled after the host is set as the starting site, that is fine since the redirection chain ends there. Let me know if I'm missing something.

Reply:
> We are only blocking requests for top-level navigations. When RenderFrameHostChanged is invoked we check if the old/new host are parent frames.

Basically, I see in `DidStartNavigation` that we overwrite starting URL of a main frame for each (non-redirect) navigation originated from any frame. Is it what we really want? `DidStartNavigation` is fired for any frame.

Reply:
> No

I think this could actually happen

Reply:
> > If the navigation is cancelled after the host is set as the starting site, that is fine since the redirection chain ends there.
> 
> sorry, didn't get it. What if the canceled navigation is not a redirect? Just a normal user-triggered navigation
>

@iefremov - If a navigation is cancelled, the redirect chain will end right. If the redirect chain ends, we don't need to worry about the `StartingSite`? I'm not able to see an edge case that we will miss with this. Do you think setting starting_site in `DidFinishNavigation` when the navigation is committed is better?

Reply:
> > We are only blocking requests for top-level navigations. When RenderFrameHostChanged is invoked we check if the old/new host are parent frames.
> 
> Basically, I see in `DidStartNavigation` that we overwrite starting URL of a main frame for each (non-redirect) navigation originated from any frame. Is it what we really want? `DidStartNavigation` is fired for any frame.

Ah! yeah. Good catch. This should be fixed now.

Reply:
> > No
> 
> I think this could actually happen

@iefremov - Is the scenario here, site A redirects to site B which is a tracker. Site B has a service worker which can interact with the indexedDB?

Reply:
> Ah! yeah. Good catch. This should be fixed now.
Maybe add a test against this scenario?

Reply:
Regarding request cancellation, consider the following scenario (completely missing any redirects):
1) Loaded page does `window.location = "https://smth"`
2) We record `smth` as a `starting_site`
3) Page does `window.stop`
4) now we didn't load `smth`, but it is recorded as a starting_site. Any requests to storage from js/service workers/whatever will be checked against the wrong entry.

Does it make sense?

Reply:
Still want to understand what happens in the scenario in my previous comment. Is that possible?

Reply:
Resolved: https://github.com/brave/brave-core/pull/403/files#diff-9a212436a82f705733d46885d21e6065R33

Reply:
> Resolved: https://github.com/brave/brave-core/pull/403/files#diff-9a212436a82f705733d46885d21e6065R33

Cool! Is there a test case for this?